refactor: Use std::span over Span

maflcko opened the pull request:
`Span` has some issues:

* It does not support fixed-size spans, which are available through `std::span`.
* It is confusing to have it available and in use at the same time with `std::span`.
* It does not obey the standard library iterator build hardening flags. See https://github.com/bitcoin/bitcoin/issues/31272 for a discussion. For example, this allows to catch issues like the one fixed in commit fabeca3458b38a3d8930cb0cbc866388c3f120f1.

Both types are type-safe and can even implicitly convert into each other in most contexts.

However, exclusively using `std::span` seems less confusing, so do it here with a scripted-diff.

Commit:
scripted-diff: Use std::span over Span

-BEGIN VERIFY SCRIPT-

 ren() { sed -i "s!\<$1\>!$2!g" $( git grep -l "$1" -- "./src" ":(exclude)src/span.h" ":(exclude)src/leveldb/db/log_test.cc" ) ; }

 ren Span            std::span
 ren AsBytes         std::as_bytes
 ren AsWritableBytes std::as_writable_bytes

 sed -i 's!SpanPopBack(Span!SpanPopBack(std::span!g' ./src/span.h

-END VERIFY SCRIPT-

Commit:
refactor: Return std::span from MakeByteSpan

In theory this commit should only touch the span.h header, because
std::span can implicilty convert into Span in most places, if needed.

However, at least when using the clang compiler, there are some
false-positive lifetimebound warnings and some implicit conversions can
not be resolved.

Thus, this refactoring commit also changed the affected places to
replace Span with std::span.

Commit:
refactor: Return std::span from MakeUCharSpan

This is possible and safe, because std::span can implicitly convert into
Span, if needed.

l0rinc commented on src/test/span_tests.cpp:
```diff
@@ -64,7 +64,7 @@ BOOST_AUTO_TEST_CASE(span_constructor_sfinae)
-    BOOST_CHECK(Spannable(Span<int>{}));
+    BOOST_CHECK(Spannable(std::span<int>{}));
     BOOST_CHECK(Spannable("char array"));
     BOOST_CHECK(Spannable(SpannableYes{}));
```
It seems to me that `SpannableYes` needs a `begin`/`end` now, not a `data`/`size` (`SpannableNo` should also be updated accordingly)
```C++
struct SpannableYes
{
    const char* begin();
    const char* end();
};
```

l0rinc reviewed (commented):
Concept ACK

Thank you for doing these big-bang cleanups!

I understand it's still a draft, but consider the following code parts that still need to be touched:
* https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md?plain=1#L859-L866
* https://github.com/bitcoin/bitcoin/blob/master/src/span.h#L32-L244
* https://github.com/bitcoin/bitcoin/blob/master/src/span.h#L257-L267

l0rinc commented:
Can `FuzzedDataProvider` and `CHMAC_SHA256` take a `const std::span<const uint8_t>` parameter now instead of separate data and size - or is that out of scope?

Commit:
refactor: Make Span an alias of std::span

This uses a macro, which can be a bit more brittle than an alias
template. However, class template argument deduction for alias templates
is only implemented in clang-19.

Commit:
refactor: Remove unused Span alias

Also, fixup some wording.

Commit:
test: Fix broken span_tests

* The comment is wrong claiming that void* was returned when void was
  returned in reality.
* The namespace is missing a name, leading to compile errors that are
  suppressed with non-standard pragmas, and leading to compile errors in
  future commits. Instead of using more non-standard suppressions, just
  add the missing name.
* The SpanableYes/No types are missing begin/end iterators, which will
  be needed when using std::span.

vasild commented:
Concept ACK

Commit:
scripted-diff: Bump copyright headers after std::span changes

Historically, the headers have been bumped some time after a file has
been touched. Do it now to avoid having to touch them again in the
future for that reason.

-BEGIN VERIFY SCRIPT-
 sed -i --regexp-extended 's;( 20[0-2][0-9])(-20[0-2][0-9])? The Bitcoin Core developers;\1-present The Bitcoin Core developers;g' $( git show --pretty="" --name-only HEAD~1 )
-END VERIFY SCRIPT-

fanquake commented:
Rebase now that we've got leveldb + the sha changes in?

Commit:
refactor: Avoid false-positive gcc warning

GCC 14.2.1 will complain about a dangling reference after replacing Span
wiht std::span. This is a false-positive, because std::find does not
return a reference.

Remove the `&` to silence the warning. Also use ranges::find while
touching the line.

src/i2p.cpp:312:21: error: possibly dangling reference to a temporary [-Werror=dangling-reference]
  312 |         const auto& pos = std::find(kv.begin(), kv.end(), '=');
      |                     ^~~
src/i2p.cpp:312:36: note: the temporary was destroyed at the end of the full expression ‘std::find<__gnu_cxx::__normal_iterator<const char*, span<const char> >, char>((& kv)->std::span<const char>::begin(), (& kv)->std::span<const char>::end(), '=')’
  312 |         const auto& pos = std::find(kv.begin(), kv.end(), '=');
      |                           ~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors

theuni commented on src/span.h:
```diff
@@ -290,9 +290,10 @@ template <typename B>
 /** Like the Span constructor, but for (const) unsigned char member types only. Only works for (un)signed char containers. */
-template <typename V> constexpr auto MakeUCharSpan(V&& v) -> decltype(UCharSpanCast(Span{std::forward<V>(v)})) { return UCharSpanCast(Span{std::forward<V>(v)}); }
+template <typename V> constexpr auto MakeUCharSpan(const V& v) -> decltype(UCharSpanCast(std::span{v})) { return UCharSpanCast(std::span{v}); }
+template <typename V> constexpr auto MakeWritableUCharSpan(V&& v) -> decltype(UCharSpanCast(std::span{std::forward<V>(v)})) { return UCharSpanCast(std::span{std::forward<V>(v)}); }
```
Unused?

theuni replied:
No preference, whatever's easiest here.

theuni reviewed (approved):
I can't find anything to complain about. It's well-scoped and the scripted-diff does the heavy lifting.

Need to fixup the headers to use `<span>` directly when possible, but that makes sense as a follow-up.

I'm looking forward to playing with static extents :)

ACK fa6f856d7146d37d51130f249ac9be9c36f7a8ad

theuni commented:
Extent-friendly version of `SpanPopBack` if you're interested. Mostly I was just curious to play around a bit.

```c++
template <typename T, size_t N = std::dynamic_extent>
T& SpanPopBack(std::span<T, N>& span)
{
    T& back = span.back();
    if constexpr (N == std::dynamic_extent)
        span = span.first(span.size() - 1);
    else
        span = span.template first<N - 1>();
    return back;
}
```

l0rinc commented on src/bench/load_external.cpp:
```diff
@@ -46,7 +46,7 @@ static void LoadExternalBlockFile(benchmark::Bench& bench)
     ss << static_cast<uint32_t>(benchmark::data::block413567.size());
     // We can't use the streaming serialization (ss << benchmark::data::block413567)
     // because that first writes a compact size.
-    ss << Span{benchmark::data::block413567};
+    ss << std::span{benchmark::data::block413567};
```
Previously
```C++
ss << Span{benchmark::data::block413567};
```
called:
```C++
template <typename Stream, BasicByte B> void Serialize(Stream& s, Span<B> span) { s.write(AsBytes(span)); }
```

But after the change it calls:
```C++
template <typename Stream, BasicByte B, std::size_t N> void Serialize(Stream& s, std::span<B, N> span) { s.write(std::as_bytes(span)); } //
```

The result in this case seems to be the same, so we might as well remove the extra span around the spans now:
```suggestion
    ss << benchmark::data::block413567;
```

l0rinc replied:
> Previously, the data was contained in a vector

A span is generated now:

```C++
inline constexpr std::span block413567{detail_block413567_raw};
```
So I don't see the reason for wrapping it in another span now.

l0rinc commented on src/test/span_tests.cpp:
```diff
@@ -42,9 +42,9 @@ using namespace spannable;
+// std::span constructor overloads that work, and disable calls to constructor overloads that
+// don't work. This makes it is possible to use the std::span constructor in a SFINAE
 // contexts like in the Spannable function above to detect whether types are or
 // aren't compatible with Spans at compile time.
```
Seems like a leftover `Spans`

l0rinc replied:
Done, you can resolve it (except that "[As a general rule, we never use an apostrophe in writing plural forms](https://www.sussex.ac.uk/informatics/punctuation/apostrophe/plurals)")

l0rinc commented on src/script/interpreter.h:
```diff
@@ -335,13 +335,13 @@ class DeferringSignatureChecker : public BaseSignatureChecker
-uint256 ComputeTapleafHash(uint8_t leaf_version, Span<const unsigned char> script);
+uint256 ComputeTapleafHash(uint8_t leaf_version, std::span<const unsigned char> script);
 /** Compute the BIP341 tapbranch hash from two branches.
   * Spans must be 32 bytes each. */
```
should this be `std::span` as well?

l0rinc replied:
Thanks, please resolve it

l0rinc reviewed (commented):
I have reviewed until fa76d5aaba34928cf403b5e142705fd03e26553f - mostly nits, only concern is a possible slight [behavior change](https://github.com/bitcoin/bitcoin/pull/31519#discussion_r1931248186).

Could you please check if the serialization method changes that resulted from this are expected (likely applied to the array-backed spans with fixed sizes only)?

Commit:
bench: Update span-serialize comment

Commit faecca9a85c1c1917d024f55cca34d19cc94f3b9 changed the type of
block413567 from vector to span, but forgot to update the comment. Do it
now.

l0rinc commented on src/span.h:
```diff
@@ -11,49 +11,49 @@
 #include <utility>
 
-/** A Span is an object that can refer to a contiguous sequence of objects.
+/** A std::span is an object that can refer to a contiguous sequence of objects.
```
shouldn't these go to the [developer notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md?plain=1#L859-L863) now? We don't have files for other std types locally which explain their behavior.

l0rinc replied:
I find the suggestions helpful, I'm fine with moving them to doc separately.
The `span's` should likely be adjusted here as well, maybe to `instances of std::span` or something similar.

l0rinc commented on src/base58.cpp:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2014-2022 The Bitcoin Core developers
```
While I find it a good idea to finally get rid of these, I think it may be cleaner to do these in a separate PR, this one's quite heavy already

l0rinc reviewed (commented):
Docker seems to complain:
> You have reached your pull rate limit

But locally all tests were passing for me.

I think this is an important change, I'd just a like a few minor changes (like remaining "Spans" references), and I'd prefer doing the unrelated header changes in a different PR for transparency.

l0rinc commented on src/test/span_tests.cpp:
```diff
@@ -24,47 +24,36 @@ bool Spannable(Ignore)
-// don't work. This makes it is possible to use the Span constructor in a SFINAE
+// Make sure template std::span template deduction guides accurately enable calls to
+// std::span constructor overloads that work, and disable calls to constructor overloads that
+// don't work. This makes it is possible to use the std::span constructor in a SFINAE
```
```suggestion
// don't work. This makes it possible to use the std::span constructor in a SFINAE
```

l0rinc reviewed (approved):
I have recreated the changes locally and compared it against your change and reviewed the diffs one-by-one.
Based on my understanding, your changes are accurate, nicely transitioning to the new type in careful commits, you have thought of every corner case.

Rebased the change against latest master, all tests are passing locally as well and I'm getting the same scripted diff results (rewritten slightly for mac).

<details>
<summary>Span->std::span</summary>

```bash
gsed -i "s#\<Span\>#std::span#g" $(git grep -l "Span" -- "./src" ":(exclude)src/span.h" ":(exclude)src/leveldb")
gsed -i "s#\<AsBytes\>#std::as_bytes#g" $(git grep -l "AsBytes" -- "./src" ":(exclude)src/span.h" ":(exclude)src/leveldb")
gsed -i "s#\<AsWritableBytes\>#std::as_writable_bytes#g" $(git grep -l "AsWritableBytes" -- "./src" ":(exclude)src/span.h" ":(exclude)src/leveldb")
gsed -i 's#SpanPopBack(Span#SpanPopBack(std::span#g' ./src/span.h
```

</details>
<details>
<summary>Copyrights</summary>

```bash
gsed -i --regexp-extended 's#( 20[0-2][0-9])(-20[0-2][0-9])? (The Bitcoin Core developers)#\1-present \3#g' $(git diff --name-only HEAD~6..HEAD)
```

</details>


I only have leftover nits (e.g. [`ss << benchmark::data::block413567;` instead of `ss << std::spam{benchmark::data::block413567};`](https://github.com/bitcoin/bitcoin/pull/31519#discussion_r1941956974) or excluding the whole `leveldb` folder in the scripted diff instead of just `":(exclude)src/leveldb/db/log_test.cc"`), but nothing critical.

Thanks for taking care of these refactorings!

ACK fa04eec5adc15f3e236e881fefd2376ce3ddc6a0

Sjors commented:
Concept ACK. I'll need to rebase some things, but I'm favor of getting that over with.

theuni commented:
@fanquake do you have an opinion on this for v29? Getting it in would make future backports easier. Could also see waiting until after branch-off to be less disruptive though.

l0rinc commented:
Note that after https://github.com/bitcoin/bitcoin/pull/30205/files#diff-aaace4d9302b33227cc9106d59234f6ae0a8946cf23fc439b3936444375d986fR282 this now needs a rebase and fix.

l0rinc commented:
ACK fab51b26d68b37f5914314ad29eeb930538ea7ae

<details>
<summary>Diff</summary>

Compared to previously acked fa77685ba5c7185781acca04f57399cdcd19e9f7, `net.cpp` and `pcp_tests.cpp` had header & Span -> std::span changes only.

</details>

laanwj commented:
Code review ACK fab51b26d68b37f5914314ad29eeb930538ea7ae
- Concept ACK: Prefer equivalent built-in functionality where possible
- Manually verified the scripted diffs
- Checked that there are no remaining unnecessary capitalizations of Span in comments.

theuni commented on src/span.h:
```diff
@@ -290,9 +290,10 @@ template <typename B>
 
 // Helper function to safely convert a Span to a Span<[const] unsigned char>.
-template <typename T> constexpr auto UCharSpanCast(Span<T> s) -> Span<typename std::remove_pointer<decltype(UCharCast(s.data()))>::type> { return {UCharCast(s.data()), s.size()}; }
+template <typename T, size_t N> constexpr auto UCharSpanCast(std::span<T, N> s) -> std::span<std::remove_pointer_t<decltype(UCharCast(s.data()))>> { return {UCharCast(s.data()), s.size()}; }
```
The resulting type here is missing the `N` parameter, which prevents it from forwarding extent info. That constructor is explicit, so we have to name the type.

With that done, there's a single case of a span being reused with an incompatible extent size. Full necessary change is:

```patch
diff --git a/src/span.h b/src/span.h
index 62e035d2d35..53eba377954 100644
--- a/src/span.h
+++ b/src/span.h
@@ -105,7 +105,9 @@ template <typename B>
 concept BasicByte = requires { UCharCast(std::span<B>{}.data()); };

 // Helper function to safely convert a span to a span<[const] unsigned char>.
-template <typename T, size_t N> constexpr auto UCharSpanCast(std::span<T, N> s) -> std::span<std::remove_pointer_t<decltype(UCharCast(s.data()))>> { return {UCharCast(s.data()), s.size()}; }
+template <typename T, size_t N> constexpr auto UCharSpanCast(std::span<T, N> s) -> std::span<std::remove_pointer_t<decltype(UCharCast(s.data()))>, N> {
+    return std::span<std::remove_pointer_t<decltype(UCharCast(s.data()))>, N>{UCharCast(s.data()), s.size()};
+}

 /** Like the std::span constructor, but for (const) unsigned char member types only. Only works for (un)signed char containers. */
 template <typename V> constexpr auto MakeUCharSpan(const V& v) -> decltype(UCharSpanCast(std::span{v})) { return UCharSpanCast(std::span{v}); }
diff --git a/src/test/util_tests.cpp b/src/test/util_tests.cpp
index bc50bf408eb..0fa6f7d8e5a 100644
--- a/src/test/util_tests.cpp
+++ b/src/test/util_tests.cpp
@@ -160,8 +160,8 @@ BOOST_AUTO_TEST_CASE(parse_hex)
     BOOST_CHECK_EQUAL_COLLECTIONS(hex_literal_span.begin(), hex_literal_span.end(), expected.begin(), expected.end());

     const std::vector<std::byte> hex_literal_vector{operator""_hex_v<util::detail::Hex(HEX_PARSE_INPUT)>()};
-    hex_literal_span = MakeUCharSpan(hex_literal_vector);
-    BOOST_CHECK_EQUAL_COLLECTIONS(hex_literal_span.begin(), hex_literal_span.end(), expected.begin(), expected.end());
+    auto hex_literal_vec_span = MakeUCharSpan(hex_literal_vector);
+    BOOST_CHECK_EQUAL_COLLECTIONS(hex_literal_vec_span.begin(), hex_literal_vec_span.end(), expected.begin(), expected.end());

     constexpr std::array<uint8_t, 65> hex_literal_array_uint8{operator""_hex_u8<util::detail::Hex(HEX_PARSE_INPUT)>()};
     BOOST_CHECK_EQUAL_COLLECTIONS(hex_literal_array_uint8.begin(), hex_literal_array_uint8.end(), expected.begin(), expected.end());
```

Arguably this is a bug introduced in this PR, as all the other conversion functions forward the extent info correctly.

theuni replied:
Noticed this when hooking up some more static extent features in a test branch btw, so not purely theoretical.

l0rinc replied:
I also noticed some inconsistencies while trying to rebase https://github.com/bitcoin/bitcoin/pull/31868#discussion_r1956480589 on top of this but didn't get to investigating why the size doesn't always propagate - thanks for the patch.

l0rinc reviewed (commented):
Found a few more static extent opportunities, but I personally don't mind doing these in separate PRs either.

l0rinc commented:
Rebased it against latest [master](https://github.com/bitcoin/bitcoin/commit/db2c57ae9eebdb75c58cd165ac929919969c19a9), ran all tests locally, searched for remaining `\bSpan\b` and compared to previously acked fab51b26d68b37f5914314ad29eeb930538ea7ae the only diff was @theuni's [comment](https://github.com/bitcoin/bitcoin/pull/31519#discussion_r1980199782).

<details>
<summary>Details</summary>

```diff
---
diff --git a/src/test/util_tests.cpp b/src/test/util_tests.cpp
--- a/src/test/util_tests.cpp(revision dc8b8e9e3340daf058a9f00afb66550cfe4f70b1)
+++ b/src/test/util_tests.cpp(date 1742215795870)
@@ -160,8 +160,8 @@
     BOOST_CHECK_EQUAL_COLLECTIONS(hex_literal_span.begin(), hex_literal_span.end(), expected.begin(), expected.end());
 
     const std::vector<std::byte> hex_literal_vector{operator""_hex_v<util::detail::Hex(HEX_PARSE_INPUT)>()};
-    hex_literal_span = MakeUCharSpan(hex_literal_vector);
-    BOOST_CHECK_EQUAL_COLLECTIONS(hex_literal_span.begin(), hex_literal_span.end(), expected.begin(), expected.end());
+    auto hex_literal_vec_span = MakeUCharSpan(hex_literal_vector);
+    BOOST_CHECK_EQUAL_COLLECTIONS(hex_literal_vec_span.begin(), hex_literal_vec_span.end(), expected.begin(), expected.end());
 
     constexpr std::array<uint8_t, 65> hex_literal_array_uint8{operator""_hex_u8<util::detail::Hex(HEX_PARSE_INPUT)>()};
     BOOST_CHECK_EQUAL_COLLECTIONS(hex_literal_array_uint8.begin(), hex_literal_array_uint8.end(), expected.begin(), expected.end());
Index: src/span.h
===================================================================
diff --git a/src/span.h b/src/span.h
--- a/src/span.h(revision dc8b8e9e3340daf058a9f00afb66550cfe4f70b1)
+++ b/src/span.h(date 1742215795870)
@@ -105,7 +105,7 @@
 concept BasicByte = requires { UCharCast(std::span<B>{}.data()); };
 
 // Helper function to safely convert a span to a span<[const] unsigned char>.
-template <typename T, size_t N> constexpr auto UCharSpanCast(std::span<T, N> s) -> std::span<std::remove_pointer_t<decltype(UCharCast(s.data()))>> { return {UCharCast(s.data()), s.size()}; }
+template <typename T, size_t N> constexpr auto UCharSpanCast(std::span<T, N> s) { return std::span<std::remove_pointer_t<decltype(UCharCast(s.data()))>, N>{UCharCast(s.data()), s.size()}; }
 
 /** Like the std::span constructor, but for (const) unsigned char member types only. Only works for (un)signed char containers. */
 template <typename V> constexpr auto MakeUCharSpan(const V& v) -> decltype(UCharSpanCast(std::span{v})) { return UCharSpanCast(std::span{v}); }

```

</details>

The remaining static extent changes can be done in follow-ups.
Let's get this merged! \\:D/

reACK ffff4a293ad878494e12f8f00108cc99ee2b713e

theuni reviewed (approved):
Thanks for addressing my comment.

I agree, this will be a fun one to get merged. I have several local branches with fun changes on top of this one.

ACK ffff4a293ad878494e12f8f00108cc99ee2b713e.
